grub: add support for passwords

buckley310 opened the pull request:
This patch adds support for user accounts/passwords in GRUB 2.
When configured, everything but the default boot option in GRUB is password-protected.

<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
This feature adds a small layer of physical security to NixOS systems.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

buckley310 committed:
grub: add support for passwords

This patch adds support for user accounts/passwords in GRUB 2.
When configured, everything but the default option is password-protected.

JohnAZoidberg commented on nixos/modules/system/boot/loader/grub/grub.nix:
```diff
@@ -137,6 +138,56 @@ in
+        };
+        description = ''
+          User accounts for GRUB. When specified, the GRUB command line and
+          all boot options except the default are password-protected.
```
It's probably good to add a warning that the passwords and hashes will be stored in the `/nix/store` world readable. And in `/boot/grub/grub.cfg`, I think.

marzipankaiser commented:
Works for me

JohnAZoidberg commented:
@0x4A6F please state what testing you have done and on what basis you are approving the PR.

0x4A6F commented:
> @0x4A6F please state what testing you have done and on what basis you are approving the PR.

I've provisioned a system with `<nixpkgs>=this pr` and used some boilerplate with following config included.

```
# ./include/pr/pr-65231.nix
{ config, pkgs, ... }:
  
# nix run nixpkgs.grub2
# grub-mkpasswd-pbkdf2

{

  boot.loader.grub.users = {
    root = { password_pbkdf2 = "grub.pbkdf2.sha512.10000.redacted"; };
  };

}
```

- [x] booting default entry without `./include/pr/pr-65231.nix`
- [x] booting default entry with `./include/pr/pr-65231.nix`
- [x] denied access to available generations and editing cmdline (with wrong user or wrong password)
- [x] accessing available generations (after input of user and valid password)
- [x] editing cmdline of default entry  (after input of user and valid password)

@JohnAZoidberg, if you could come up with other scenarios, I'm eager to test.

buckley310 commented on nixos/modules/system/boot/loader/grub/grub.nix:
```diff
@@ -137,6 +138,67 @@ in
+                This hash will be copied to the Nix store, and will be visible to all local users.
+              '';
+            };
+            password_plaintext_file = mkOption {
```
In retrospect, I wonder if the ability to specify credentials inline is unnecessary. It's extra code paths to test, and it's possible nobody would use it (I probably wouldn't).

Additionally, the perl script can just check if hashes start with "grub.pbkdf2". This would leave room for future hash formats to be supported under the same single option, and get rid of the ugly "pbkdf2" in the option name.

That would leave `passwordFile` and `hashedPasswordFile`, which sounds fine to me.

rnhmjoj replied:
> In retrospect, I wonder if the ability to specify credentials inline is unnecessary

Security-wise these options should be avoided but they are convenient sometimes: in NixOS tests for examples.

> Additionally, the perl script can just check if hashes start with "grub.pbkdf2"...

This looks like a good idea.

buckley310 replied:
Changes added. When I get a chance to spin up an VM I'll test everything.

buckley310 commented:
I have just tested with the following config. Of the 5 "boot.loader" lines, I tested them one at a time with the others commented out.
```
{ config, pkgs, ... }:
let
  pw = "zzz";
  pwh = "grub.pbkdf2.sha512.10000.BA1BE1 ...[snip]... D10FBBF04CF4";
in {

  boot.loader.grub.users.root.password = pw;
  boot.loader.grub.users.root.passwordFile = "${pkgs.writeText "pw" pw}";
  boot.loader.grub.users.root.hashedPassword = pwh;
  boot.loader.grub.users.root.hashedPasswordFile = "${pkgs.writeText "pwh" pwh}";
  boot.loader.grub.users.root.hashedPassword = "not a valid hash";

  ...[snip]...
}
```

rnhmjoj commented:
I tested every options and it's working as expected. The only issue I've found is that `nixos-rebuild build-vm-with-bootloader` fails with no meaningful if the hash or the password file is invalid.

I guess if you `nixos-rebuild switch` your system the error will be explicit but I have yet to test it.

rnhmjoj commented:
ping

prusnak commented:
Fixes https://github.com/NixOS/nixpkgs/issues/20272

buckley310 commented:
Is there a good way to emit an error during a buold-vm event?

rnhmjoj commented:
I think what you are doing is just fine, the problem must be `build-vm-with-bootloader` is ignoring failures in the install script. I tested with nixos-install and it does work:

```
installing the boot loader...
setting up /etc...
updating GRUB 2 menu...
Can't read file 'doesnotexist.pwd'! at /nix/store/qps12bpls270qk8pb298abjzsjvnb00w-install-grub.pl line 264.
```

It would be good to have a NixOS installer test for grub with passwords but I get the OCR to work in the grub screen. I'll try some more before giving up.
Otherwise I'll go on and merge this as is.

rnhmjoj committed:
nixos/lib/test-driver: add wait_for_console_text

This method is similar to wait_for_text but is based on matching
serial console lines instead of the VGA output.

rnhmjoj committed:
nixos/tests: add test for grub authentication

rnhmjoj commented:
So, I managed to get the output of grub to the serial console, which is easier to deal with.
I added a test to check if the password protection and booting works.

rnhmjoj commented:
@tfc can you take a look at the test driver part?

tfc commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -738,10 +759,19 @@ def create_socket(path: str) -> socket.socket:
 
+        # Store last serial console lines for use
+        # of wait_for_console_text
+        self.last_lines: Queue = Queue(maxsize=1000)
```
Initializing this queue with a finite buffer brings the risk that this might make the script block (we had that already once). When we make the buffer size infinite, the risk that it gets too large is negligible.

reference: https://docs.python.org/3/library/queue.html#queue.Queue

rnhmjoj replied:
Well, this is intended to store a few "screens", which are approximately 100 lines each. 1000 lines are 10 "screens", which is already quite a lot if you are waiting for something to appear, and you have to consider that `wait_for_console_text` is constantly reading lines. What value do you propose?

tfc commented on nixos/tests/grub.nix:
```diff
@@ -0,0 +1,52 @@
+        machine.send_chars("alice\n")
+        machine.wait_for_console_text("Enter password:")
+        machine.send_chars("supersecret\n\n")
+        machine.wait_for_console_text("Linux version")
```
wrapping these repeating lines into a function so that you can say

```python
select_all_configurations()
with subtest("Invalid credentials are rejected"):
        grub_login_as("wronguser", "wrongsecret")

select_all_configurations()
with subtest("Valid credentials are accepted"):
    grub_login_as("alice", "supersecret")
```

would make this test a nice example of how declarative integration test can be looking.

rnhmjoj replied:
Good idea, I will to it this way.

tfc commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -675,6 +676,26 @@ def screen_matches(last: bool) -> bool:
         with self.nested("waiting for {} to appear on screen".format(regex)):
             retry(screen_matches)
 
+    def wait_for_console_text(self, regex: str) -> None:
```
can you please document this and the OCR-using function in a way that the user understands that both basically do the same but also both exist for good reasons?

tfc replied:
also, in the other function i could match over lines while this one matches per-line. this is something that we either want to unify or document.

rnhmjoj replied:
I added a small description in `writing-nixos-test.xml`, should I add a comment in the script or document it somewhere else?

rnhmjoj replied:
> also, in the other function i could match over lines while this one matches per-line. this is something that we either want to unify or document.

Right, it's better to keep the two equal. I'll see what I can do to make it work on a "screen".

tfc commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -675,6 +676,26 @@ def screen_matches(last: bool) -> bool:
+                text = "\n> ".join(screen)
+                screen = []
+                self.log(
+                    "Last attempt failed. Text was:\n===\n> {}\n===\n".format(text)
```
assuming i have 1000 lines in the buffer, this would fill 10 screens. if i excpect my keyword to reside very late in one of the last 1000 lines, then i will get 9 log lines that say "last attempt failed", although i would not consider this an error. Or am i completely off here?

rnhmjoj replied:
That's correct. Do you think it's too verbose? I copied the approach of `wait_for_text` but it may not really appropriate in this case.

rnhmjoj replied:
Dropped.

tfc commented on nixos/lib/test-driver/test-driver.py:
```diff
@@ -738,10 +759,19 @@ def create_socket(path: str) -> socket.socket:
                 line = _line.decode(errors="ignore").replace("\r", "").rstrip()
+                try:
+                    self.last_lines.put(line, block=False)
+                except queue.Full:
```
i would rather see first that we really have the problem that some test fails because it goes OOM or gets too slow processing serial output, before we begin dealing with silently failing tests because we drop serial lines without further notice because a small 1000-line buffer overruns.

rnhmjoj replied:
Sorry, I don't what you're asking here. Do you want this operation to be blocking or are you proposing to increase the buffer size (as above)?

tfc replied:
i think that it would be a good idea to:

- make the buffer size infinite (get that with empty Queue constructor args)
- make the `put` calls blocking, assuming the queue never runs full
- revisit the question if it is really necessary to model the notion of a 100-lines "screen", because it makes a difference for the caller if the "screen buffer" has 1, 10, or 100 screens stored currently. there is no way to get to the "last" screen. wouldn't it suffice to give `wait_for_console_text` semantics that are like "does the serial output match my regex since boot/i called this last time"?

rnhmjoj replied:
Ok, I removed the size limit and stored the lines received since `wait_for_console_text` is called in a `StringIO`.
This way we efficiently append lines into the buffer and can match multiline regexes.

rnhmjoj replied:
ping @tfc

tfc replied:
LGTM!

rnhmjoj commented:
Thank you for the review!
